src/main/scala/calc/Typer.scala

```diff
@@ -0,0 +1,25 @@
+package calc
+
+object Typer {
+
+  def inferType(tree: Tree): TreeT = {
+    tree match {
+      case t:Literal => implicit val pos = t.pos
+        LiteralT(t.value)
+
+      case t:BinaryOp => binaryOp(t)
+    }
+  }
+
+  def binaryOp(t: BinaryOp) = { implicit val pos = t.pos
+    val lhs = inferType(t.lhs)
+    val rhs = inferType(t.rhs)
+    if (lhs.getType() == rhs.getType()) {
```

Comment:
That's not strict enough. A binary op demands that both sides be doubles. So if both sides are functions, it's still a type error.